botorch/utils/sample_pareto.py

```diff
@@ -0,0 +1,281 @@
+
+    Args:
+        model: The random Fourier feature GP sample.
+        bounds: A `2 x d`-dim Tensor containing the input bounds.
+        num_objectives: The number of objectives.
+        num_generations: The number of generations of NSGA2.
+        pop_size: The population size maintained at each step of NSGA2.
+        num_offsprings: The number of offsprings used in NSGA2.
+        maximize: If true we solve for the Pareto maximum.
+
+    Returns:
+        A two-element tuple containing
+
+        - pareto_sets: A `num_pareto_points x d`-dim Tensor containing the Pareto
+            optimal set of inputs.
+        - pareto_fonts: A `num_pareto_points x num_objectives`-dim Tensor containing
+            the Pareto optimal set of objectives.
+    """
+    tkwargs = {"dtype": bounds.dtype, "device": bounds.device}
+    d = bounds.shape[-1]
+    weight = -1.0 if maximize else 1.0
+
+    class PymooProblem(Problem):
+        def __init__(self):
+            super().__init__(
+                n_var=d,
+                n_obj=num_objectives,
+                n_constr=0,
+                xl=bounds[0].cpu().detach().numpy(),
+                xu=bounds[1].cpu().detach().numpy(),
+            )
+
+        def _evaluate(self, x, out, *args, **kwargs):
+            xt = torch.tensor(x, **tkwargs)
+            out["F"] = weight * model.posterior(xt).mean.cpu().detach().numpy()
+            return out
+
+    # Use NSGA2 to generate a number of Pareto optimal points.
+    results = minimize(
+        problem=PymooProblem(),
+        algorithm=NSGA2(
+            pop_size=pop_size,
+            n_offsprings=num_offsprings,
+            sampling=FloatRandomSampling(),
+            crossover=SBX(prob=0.9, eta=15),
+            mutation=PolynomialMutation(eta=20),
+            eliminate_duplicates=True,
+        ),
+        termination=get_termination("n_gen", num_generations),
+    )
+
+    if model.num_outputs == 1:
+        ps = torch.tensor(results.X, **tkwargs).unsqueeze(0)
+        pf = weight * torch.tensor(results.F, **tkwargs).unsqueeze(0)
+    else:
+        ps = torch.tensor(results.X, **tkwargs)
+        pf = weight * torch.tensor(results.F, **tkwargs)
+
+    pareto_mask = is_non_dominated(-1.0 * weight * pf)
+    pareto_set = ps[pareto_mask]
+    pareto_front = pf[pareto_mask]
+
+    return pareto_set, pareto_front
+
+
+def sample_pareto_sets_and_fronts(
+    model: Model,
+    bounds: Tensor,
+    num_pareto_samples: int,
+    num_pareto_points: int,
+    maximize: bool = True,
+    num_generations: int = 100,
+    pop_size: int = 100,
+    num_offsprings: int = 10,
+    num_rff_features: int = 512,
+    max_tries: int = 3,
+    num_greedy: int = 0,
+    X_baseline: Optional[Tensor] = None,
+    ref_point: Optional[Tensor] = None,
+) -> Tuple[Tensor, Tensor]:
+    r"""Computes the Pareto optimal set and front from samples of the GP.
+
+    (i) Samples are generated using random Fourier features.
```

Comment:
This need not be RFFs. It could also use decoupled sampling. Let's add a TODO to generalize the GP sampling bit